src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -801,6 +543,314 @@ private void BlobLeaseManagerHasLeaseChanged(object sender, EventArgs e)
             IsPrimaryChanged?.Invoke(this, EventArgs.Empty);
         }
 
+        /// <summary>
+        /// Perform any early initialization operations.
+        /// </summary>
+        private void Preinitialize()
+        {
+            string debugSentinelFileName = Path.Combine(_hostLogPath, ScriptConstants.DebugSentinelFileName);
+            LastDebugNotify = File.GetLastWriteTime(debugSentinelFileName);
+
+            // take a startup time function directory snapshot so we can detect function additions/removals
+            // we'll also use this snapshot when reading function metadata as part of startup
+            // taking this snapshot once and reusing at various points during initialization allows us to
+            // minimize disk opoerations
+            _directorySnapshot = Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).ToImmutableArray();
```

Comment:
As part of some perf investigations, I'm now taking this snapshot early and reusing it in a couple places during init, which saves us 2 directory enumerations. I haven't measured the startup improvement from that yet, but on Azure Files I'd expect it to matter.